sintr_common/lib/tasks.dart

```diff
@@ -0,0 +1,315 @@
+// Copyright (c) 2015, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:appengine/appengine.dart' as ae;
+import 'package:gcloud/db.dart' as db;
+import 'package:memcache/memcache.dart' as mc;
+import 'package:sintr_common/gae_utils.dart';
+import 'package:sintr_common/logging_utils.dart' as logging;
+
+const int DATASTORE_TRANSACTION_SIZE = 100;
+
+db.DatastoreDB _db = ae.context.services.db;
+final _logger = new logging.Logger("tasks");
+
+/// Abstraction over a piece of work to be done by a compute node
+class Task {
+
+  // Location of the object in the datastore
+  final db.Key _objectKey;
+  _TaskModel backingstore;
+
+  String get _stateMemcacheKey => "$_objectKey-LifecycleState";
+  String get _pingMemcacheKey => "$_objectKey-LastPing";
+
+  /// Get the state of the task
+  // The state machine for READY -> ALLOCATED is synchronised
+  Future<LifecycleState> get state async {
+    _logger.finer("Get state on Task for $_objectKey");
+    String memcacheResult = await SafeMemcache.get(_stateMemcacheKey);
+
+    if (memcacheResult != null) {
+      LifecycleState result = _lifecyclefromInt(int.parse(memcacheResult));
+      _logger.finer("Get state on Task for $_objectKey -> $result: OK");
+      return result;
+    }
+
+    await _pullBackingStore();
+    if (backingstore == null) return null;
+
+    LifecycleState result = _lifecyclefromInt(backingstore.lifecycleState);
+    _logger.finer("Get state on Task for $_objectKey -> $result: OK");
+    return result;
+  }
+
+  /// Set the state of the task, do not use to set to ALLOCATED
+  Future setState(LifecycleState state) async {
+    if (state == LifecycleState.ALLOCATED)
+      throw "Setting tasks to allocated may only be done by the TaskController";
+
+    _logger.finer("Set state on Task for $_objectKey -> $state");
+
+    await _pullBackingStore();
+    int lifeCycleStateInt = _intFromLifecycle(state);
+    await SafeMemcache.set(
+      _stateMemcacheKey, lifeCycleStateInt.toString());
+    backingstore.lifecycleState = lifeCycleStateInt;
+    await _pushBackingStore();
+
+    _logger.finer("Set state on Task for $_objectKey -> $state : OK");
+  }
+
+  // Number of times this task has failed to execute
+  Future<int> get failureCounts async {
+    await _pullBackingStore();
+    return backingstore?.failureCount;
+  }
+
+  // Last time this task was pinged as having made progress
+  // Best effort syncronised
+  Future<int> get lastPingEpochTime async {
+    String memcacheResult = await SafeMemcache.get(_pingMemcacheKey);
+
+    if (memcacheResult != null) {
+      return int.parse(memcacheResult);
+    }
+
+    await _pullBackingStore();
+    if (backingstore == null) return null;
+
+    return backingstore.lastUpdateEpochMs;
```

Comment:
It might improve readability if the memcache key names (e.g. _pingMemcacheKey) and the backingstore field names (e.g. lastUpdateEpochMs) were better aligned. At first glance it is not apparent that one is the memcache value for the other... until you carefully read the code and it becomes apparent.